### PR TITLE
fix: chat auto-scroll is fragile in chromium environments

### PR DESCRIPTION
**Pull request by danjrwalsh**
Potential fix for #660 and #700

In Chromium environments (including Electron, so this affects the desktop apps too), I've found the existing auto-scroll to be quite fragile, breaking fairly often (multiple times an hour) from regular usage. Every time it happens, it requires switching chat channels or pressing `Esc` a couple times to get it functioning again. It's a minor nuisance, but it does get frustrating. I was unable to reproduce the issue on Firefox, which is why I'm assuming this is a Chromium specific issue.

To get around what appears to be a CSS issue on Chromium, I've added a scroll event listener and ResizeObserver/MutationObserver that will keep the chat history snapped to the bottom as long as the most recent message was still visible (even partially).

Live example of the two instances side-by-side (in Brave browser), mimicking actions (left prod, right local). You can see prod lost auto scroll (scrolling back down does **not** re-engage the scrolling).
stoat

NOTE: While I have tested this in a browser environment, I have **not** tested this inside a local Electron instance.

Feedback is welcome, this is my first contribution to both Stoat and a Solid.js project in general (more of a Vue/Nuxt guy).

### CODE REVIEW

File: packages/client/components/ui/components/utils/ListView2.tsx [hunk 1]
```diff
@@ -102,6 +118,34 @@ export function ListView2(props: Props) {
...
+
+    el.addEventListener("scroll", onScroll, { passive: true });
+    onCleanup(() => el.removeEventListener("scroll", onScroll));
+  });
+
+  createEffect(() => {
+    const el = contentRef;
+    if (!el) return;
+
+    const observer = new ResizeObserver(() => {
```
[the remark this answers is not in the document]
Comment (danjrwalsh) [reply packages/client/components/ui/components/utils/ListView2.tsx]: We specify column-reverse in the comment above `BOTTOM_THRESHOLD`. I'll leave it up to the maintainers to decide whether they want a secondary comment for this or not.

File: packages/client/components/ui/components/utils/ListView2.tsx [hunk 2]
```diff
@@ -58,17 +58,30 @@ interface Props {
...
 /**
  * Dynamic list view with ability to move through history
  *
  * This component only provides the scrolling behaviour
  */
 export function ListView2(props: Props) {
   let ref: HTMLDivElement | undefined;
+  let contentRef: HTMLDivElement | undefined;
+
+  let wasAtBottom = true;
```
[the remark this answers is not in the document]
Comment (danjrwalsh) [reply packages/client/components/ui/components/utils/ListView2.tsx]: I believe the app already defaults to bottom automatically, even when switching to a chat channel that was previously scrolled up. If desired by the maintainers, I can make this change, but as of right now it appears to maintain the existing functionality.

**Comment by mihaicm93** [comment]
> I've been running this locally side-by-side the current prod for the last couple hours and I've yet to experience auto scroll break on me once, whereas prod has lost auto scroll 4 times.

I have tested it both on Chrome and in the dev client and had it happen there as well, feels a lot more consistent but still had issues with it loosing scroll, you should try interact with messages,  replies and check profiles it's very random but still happens happens that it looses scroll

**Comment by danjrwalsh** [comment]
> > I've been running this locally side-by-side the current prod for the last couple hours and I've yet to experience auto scroll break on me once, whereas prod has lost auto scroll 4 times.
> 
> 
> 
> I have tested it both on Chrome and in the dev client and had it happen there as well, feels a lot more consistent but still had issues with it loosing scroll, you should try interact with messages,  replies and check profiles it's very random but still happens happens that it looses scroll

I did try a bunch of regular actions like that yesterday when using it over a period of a couple hours but I guess I just got lucky. 

Chances are it's just the threshold though, it's only set to 1px currently, increasing it would likely catch those remaining scenarios you ran into.

I'll play around with the threshold after work and see if I can find a sweet spot.

**Comment by danjrwalsh** [comment]
@mihaicm93 I didn't have much time today but tried going a slightly different approach that I think will be more reliable + provide a better UX. 

Instead of having a static threshold, it's now checking if the last message was visible (including partially) by comparing bounding rects and if so, keeping the auto-scrolling engaged. So now auto-scroll should continue working unless you explicitly move the most recent message completely out of bounds (I also added a mutation observer so it only takes into account explicit user scrolls).

Give it a test on your side when you have a moment and let me know what you think. It's a more "aggressive" way of handling it, but covers more edge cases. If you think it's a bit too aggressive and doesn't release auto-scroll early enough, I can revert back to the static threshold approach and simply increase the threshold (30-50px felt best in my testing, but that could be subjective).

**Review by mihaicm93 — COMMENTED** [review]
I've been testing it for a few hours today, desktop and browser, worked good for me without loosing track

**Review by mihaicm93 — APPROVED** [review]

File: packages/client/components/ui/components/utils/ListView2.tsx [hunk 3]
```diff
@@ -65,10 +65,32 @@ interface Props {
...
+    const messages = contentRef.querySelectorAll<HTMLElement>("[id]");
+    const newest = messages[messages.length - 1];
+    if (!newest) return false;
+
+    const containerRect = ref.getBoundingClientRect();
+    const msgRect = newest.getBoundingClientRect();
+
+    return (
+      msgRect.top < containerRect.bottom && msgRect.bottom > containerRect.top
+    );
```
Comment (insertish) [inline packages/client/components/ui/components/utils/ListView2.tsx:86]: Could this not just refer to the last element in the list container?

**Review by insertish — CHANGES_REQUESTED** [review]
I don't super follow the logic from a quick skim, but to be fair, this part of the project has always been convoluted and confusing. If this has been tested and it works, then it's not a huge issue since it's relatively self-contained.

But at least annotating what `snap`, etc. means (or better just using more verbose variable names) would go a long way.

\* please mark as ready for review afterwards!

**Comment by Dadadah** [comment]
This was fixed in #1160
